File: language/golang/verifyvpcsecurehash.go
package main

import (
	"fmt"
	"net/url"
	"strings"
)

func onePayVerifySecureHash(urlMerchant string, merchantHashCode string) {
	u, err := url.Parse(urlMerchant)
	if err != nil {
		panic(err)
	}
	queryParams := u.Query()
	queryParamsMap := make(map[string]string)
	for k, v := range queryParams {
		queryParamsMap[k] = strings.Join(v, "")
	}
	queryMapSorted := sortParams(queryParamsMap)
	stringToHash := generateStringToHash(queryMapSorted)
	onePaySecureHash := generateSecureHash(stringToHash, merchantHashCode)
	merchantSecureHash := queryParamsMap["vpc_SecureHash"]
	fmt.Println("OnePay's Hash: ", onePaySecureHash)
	fmt.Println("Merchant's Hash: ", merchantSecureHash)
	if onePaySecureHash == merchantSecureHash {
		fmt.Println("vpc_SecureHash is valid")
	} else {
		fmt.Println("vpc_SecureHash invalid")
	}

}

File: language/javascript/VerifyVpcSecureHash.js
const util = require("./Util");

function onePayVerifySecureHash(urlResponse, merHashCode) {
  let newURL = new URL(urlResponse);
  let params = new URLSearchParams(newURL.search);
  let hashFromMerchant = params.get("vpc_SecureHash");
  let paramObject = Object.fromEntries(params);
  let paramsSorted = util.sortObj(paramObject);
  let stringToHash = util.generateStringToHash(paramsSorted);
  let OnePaySign = util.genSecureHash(stringToHash, merHashCode);
  console.log("onepay's hash: " + OnePaySign);
  console.log("merchant's hash: " + hashFromMerchant);
  if (OnePaySign == hashFromMerchant) {
    console.log("vpc_SecureHash PASSED!!!");
  } else {
    console.error("invalid vpc_SecureHash");
  }
}

module.exports = {
  onePayVerifySecureHash,
};
